pkg/util/sys/device.go

```diff
@@ -399,17 +409,83 @@ func parseUdevInfo(output string) map[string]string {
 	return result
 }
 
-func isDeviceAvailable(executor exec.Executor, devicePath string) (bool, string, error) {
-	CVInventory, err := inventoryDevice(executor, devicePath)
+func isDeviceAvailable(executor exec.Executor, devicePath string) (available bool, rejectedReasons string, outErr error) {
+	output, err := executor.ExecuteCommandWithOutput("lsblk", devicePath,
+		"--bytes", "--noheadings", "--output", "RM,RO,SIZE,TYPE")
 	if err != nil {
-		return false, "", fmt.Errorf("failed to determine if the device %q is available. %v", devicePath, err)
+		return false, "", errors.Wrapf(err, "failed to determine if device %q is available", devicePath)
+	}
+	items := strings.Fields(output)
+	if len(items) != 4 {
+		return false, "", errors.Errorf(
+			"failed to determine if device %q is available. lsblk output should have 4 fields but has %d: %s", devicePath, len(items), items)
 	}
+	rm := items[0]
+	ro := items[1]
+	sizeStr := items[2]
+	devType := items[3]
 
-	if CVInventory.Available {
-		return true, "", nil
+	rejectedList := []string{}
+
+	// must not be removable
+	if rm != "0" { // 0 means not removable, 1 means removable
+		rejectedList = append(rejectedList, "removable")
+	}
+
+	// must not be read-only
+	if ro != "0" { // 0 means not read-only, 1 means read-only
+		rejectedList = append(rejectedList, "read-only")
+	}
+
+	// must be large enough
+	sizeBytes, err := strconv.Atoi(sizeStr)
+	if err != nil {
+		return false, "", errors.Wrapf(err,
+			"failed to determine if device %q is available. failed to interpret reported size %q", devicePath, sizeStr)
 	}
+	if sizeBytes < 5368709120 /* 5368709120 = 5GB */ {
+		rejectedList = append(rejectedList, "insufficient space (< 5GB)")
+	}
+
+	// must be raw device or partition
+	// rejects devType == "lvm" if is an LVM device
+	if devType != DiskType && devType != PartType {
+		rejectedList = append(rejectedList, fmt.Sprintf(`device type %q is not acceptable; should be raw device ("disk") or partition ("part")`, devType))
+	}
+
+	// must not be locked
+	// - try to open in read-write exclusive mode to test if it's locked or not
+	f, err := os.OpenFile(devicePath, os.O_RDWR|os.O_EXCL, os.FileMode(0))
```

Comment:
I'd just open with `O_RDONLY` instead. We already know the device is not read-only and I don't want to conflict with `systemd-udev` for locks.

Also, should we use `os.ModeDevice()`  instead of `os.FileMode(0)`? Feels more appropriate.

Reply:
I used the same options here that ceph-volume uses. It uses O_EXCL as well as mode 0. I suspect that using O_RDONLY wouldn't cause an error if the disk is opened exclusively by a database. We certainly could change this, but I'm not totally sure what effects it might have on trying to change it.